### DISCUSSION TITLE
README claims pep 646 is supported but...the imports don't even seem to exist in Python interpreter

### DISCUSSION THREAD

**Discussion by zzzeek** [microsoft/pyright#2780] [Q&A]
Looking at https://github.com/microsoft/pyright/commit/dfa5892b73c2119c53bb5f92a08c8584b2f42b2c, pep 646 variadic generics is claimed to be supported.

But my Python 3.10 interpreter doesn't even seem to support the imports, or the syntax, at all, and pyright doesnt recognize it either.

The program below is taken from https://www.python.org/dev/peps/pep-0646/#type-variable-tuples

```py
from typing import TypeVar, Generic, Tuple, NewType
from typing_extensions import TypeVarTuple

Shape = TypeVarTuple("Shape")

Height = NewType('Height', int)
Width = NewType('Width', int)

class Array(Generic[*Shape]):

    def __init__(self, shape: Tuple[*Shape]):
        self._shape: Tuple[*Shape] = shape

    def get_shape(self) -> Tuple[*Shape]:
        return self._shape

shape = (Height(480), Width(640))
x: Array[Height, Width] = Array(shape)

y = abs(x)  # Inferred type is Array[Height, Width]
z = x + x   #        ...    is Array[Height, Width]
```

but there's no TypeVarTuple either in "typing" or "typing_extensions":

```
$ python
Python 3.10.0 (default, Nov  5 2021, 17:23:47) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import typing
>>> typing.TypeVarTuple
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'typing' has no attribute 'TypeVarTuple'
>>> import typing_extensions
>>> typing_extensions.TypeVarTuple
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'typing_extensions' has no attribute 'TypeVarTuple'
```

the program doesnt even compile, the syntax is not supported:

```
$ python test3.py 
  File "/home/classic/dev/sqlalchemy/test3.py", line 9
    class Array(Generic[*Shape]):
               ^
SyntaxError: expected ':'

```

so...what exactly is the pep-646 support in pyright?   within pylance I also see an error like this:

```
(variable) Shape: Type[Shape@Array]
Unpack operator in subscript is not supportedPylance
```

if I change it to this:

```py
class Array(Generic[Shape]):

    def __init__(self, shape: Tuple[Shape]):
        self._shape: Tuple[Shape] = shape

```

then it says the opposite, ha! funny  "Expected unpacked TypeVarTuple; use Unpack[Shape] or *ShapePylance"

so...is it like there's some kind of pep-646 mechanism in pyright / pylance but it's basically useless right now?  im not understanding why a Python 3 pep has a syntax that the interpreter can't even accept.

**Comment by erictraut** [comment]
PEP 646 is still in "draft" form. It's open for comment and is being considered for inclusion in Python 3.11. We typically try to get ahead of type-related functionality and incorporate support for them in pyright early. This helps find holes in the standard and gives people a chance to play with the new feature even though it isn't yet implemented in the runtime.

`TypeVarTuple` and `Unpack` are not yet defined in the official typeshed stubs, but I include definitions for them in the version of the `typing_extensions.pyi` stub that ships with pyright and pylance.

The unpack operator in a subscript operator is part of a draft [PEP 637](https://www.python.org/dev/peps/pep-0637/) that was rejected for Python 3.10. The authors of PEP 646 are hoping to include the unpack operator within subscript expressions as part of PEP 646, but the `Unpack[x]` symbol is a backup in case that part of the proposal is rejected by the Python committee.

So, to answer your question, pyright has full type-checking support for the latest draft of PEP 646. You and others can use it to play around with the functionality and provide feedback on the PEP while it's still in the "accepting comments" phase.

**Reply by erictraut** [reply]
Oh, if you want to use the unpack operator syntax, it's currently enabled if you override the `pythonVersion` in pyright to `3.11`. Or you can replace `*Shape` in the sample above with `Unpack[Shape]`, and that will work without switching to `3.11`.

**Reply by zzzeek** [reply]
OK so as far as the symbols, is it sufficient if I import from typing_extensions within a TYPE_CHECKING block, or does everything have to be in a .pyi file ?

basically if I can just see that I can get the pep 646 idea working for SQLAlchemy that will be a big win for when it goes through.

**Reply by zzzeek** [reply]
OK I can get this working with some boilerplate, example below typechecks completely and also runs in Python.

the one issue is that dealing with the TypeVarType objects without knowing the length or element types seems very cumbersome.   See example below, if you have any suggestions.

```py
from typing import TypeVar, Generic, Tuple, NewType, cast
import typing

if typing.TYPE_CHECKING:
    from typing_extensions import TypeVarTuple, Unpack

else:
    TypeVarTuple = TypeVar

    class Unpack:
        def __class_getitem__(self, typeargs):
            return typeargs


Shape = TypeVarTuple("Shape")

Height = NewType("Height", int)
Width = NewType("Width", int)


class Array(Generic[Unpack[Shape]]):
    def __init__(self, shape: Tuple[Unpack[Shape]]):
        self._shape: Tuple[Unpack[Shape]] = shape

    def get_shape(self) -> Tuple[Unpack[Shape]]:
        return self._shape

    def __abs__(self) -> "Array[Unpack[Shape]]":

        # Expression of type "tuple[_T@abs, ...]" cannot be assigned to declared type "Tuple[*Shape@Array]"
        # TypeVarTuple cannot be bound to a tuple of unknown lengthPylancereportGeneralTypeIssues
        #
        # newvalue_attempt_one: Tuple[Unpack[Shape]] = tuple(
        #     [abs(elem) for elem in self._shape]
        # )

        # this one, it doesn't know "elem" is an int, because TypeVarTuple
        # can't be "bound" yet (mentioned at
        # https://www.python.org/dev/peps/pep-0646/#variance-type-constraints-and-type-bounds-not-yet-supported)
        #
        # so we get:
        #
        # Argument of type "Shape@Array" cannot be assigned to parameter "__x" of type "SupportsAbs[_T@abs]" in function "abs"
        # "object*" is incompatible with protocol "SupportsAbs[_T@abs]"
        # "__abs__" is not presentPylancereportGeneralTypeIssues
        #
        # newvalue_attempt_two: Tuple[Unpack[Shape]] = cast(
        #     Tuple[Unpack[Shape]], tuple([abs(elem) for elem in self._shape])
        # )

        # this one, with casts everywhere, this one passes.  is this
        # how it has to be?
        newvalue_attempt_three = cast(
            Tuple[Unpack[Shape]],
            tuple([abs(cast(int, elem)) for elem in self._shape]),
        )

        return Array(newvalue_attempt_three)

        # this is fine
        # return Array(self._shape)


shape = (Height(480), Width(640))
x: "Array[Height, Width]" = Array(shape)


y = abs(x)  # Inferred type is Array[Height, Width]

```

**Reply by zzzeek** [reply]
also it obviously falls over completely in mypy.  assuming mypy isn't going to have pep 646 for many months, i would need to determine what our strategy would be if we wanted to release code using this.  if there's a way to get mypy to degrade gracefully, for example.

**Reply by erictraut** [reply]
That's great to hear that you're finding the functionality in PEP 646 useful. @mrahtz has been driving the evolution of this PEP, so I'm sure he'll be interested in the feedback.

Adding support for PEP 646 in pyright was a big undertaking. For comparison, it was similar in complexity to adding support for PEP 612. Mypy still doesn't have full support for PEP 612, which was ratified two years ago. So it's quite possible that mypy's support for PEP 646 will lag by 2+ years from the time PEP 646 is ratified. Also, it's not guaranteed to that PEP 646 will be accepted in time for Python 3.11. It's still under discussion. 

If you're interested in PEP 646, @mrahtz has been holding monthly meetings (announced in the typing-sig mail group) to gather feedback and refine the proposal.

**Reply by zzzeek** [reply]
also is any of this syntax supported?  https://www.python.org/dev/peps/pep-0646/#unpacking-unbounded-tuple-types

I'm trying to substitute "Unpack" for "*" in those and it's not working:

```py
    @overload
    def execute(
        self,
        statement: "Select[Unpack[Tuple[_T, ...]]]",
        parameters: Optional[Mapping[str, Any]] = None,
        execution_options: Optional[Mapping[str, Any]] = None,
    ) -> "Result[Unpack[Tuple[_T, ...]]]":
        ...
```

errors: "Expected TypeVarTuple as type argument for Unpack"

**Reply by erictraut** [reply]
In an earlier draft of the PEP (the one that I used as the basis for the current pyright implementation), unpacking was allowed only with TypeVarTuple arguments. It appears that in a recent draft, support was added for unpacking tuples. I'll need to add that to my to-do list.

**Reply by zzzeek** [reply]
OK, I'll leave you alone for now.  happy new year!

**Reply by mrahtz** [reply]
Just to confirm, everything Eric says is correct - PEP 646 is currently still in draft, waiting for approval by the Python steering council.

Unfortunately, yes, it probably will be a while before it's supported by all type checkers. Pyre probably has the best support, and thanks to Eric's hard work Pyright support should also be excellent except for unpacking of tuples.

If you do have any feedback on the PEP while playing around with SQLAlchemy, we would indeed appreciate that feedback very much. https://github.com/mrahtz/cpython/tree/pep646 where we're working on our draft implementation in CPython might also be useful to you.

Happy new year to you both!

**Reply by zzzeek** [reply]
thanks!   I got some POCs working for pep-646 yesterday and the limitation I was coming up against I don't think is intrinsic to pep-646, it's intrinsic to pep-484 in general and I dont think it's going to change.  

For background I've illustrated how I (erroneously, I believe) had us design sqlalchemy2-stubs, and how that doesn't really work out even though it seems more correct to me, and we should be instead going with how @jukkal originally set up typing for column expressions in sqlalchemy-stubs, this is at https://github.com/sqlalchemy/sqlalchemy/issues/7519
